server/routes/api/users.js

```diff
@@ -75,4 +77,99 @@ router.post("/login", (req, res) => {
   }
 });
 
+// @route GET /api/users/{id}
+// @desc Get a User object
+// @access Authenticated User can get own record
+router.get(
+  "/:id",
+  passport.authenticate("jwt", { session: false }),
+  async (req, res) => {
+    let id = req.params.id;
+    console.log(req.user);
+
+    try {
+      if (req.user.id === id) {
+        await User.findById(id, function(err, user) {
+          if (!user) {
+            res.status(404).send({ id: `User with id ${id} is not found` });
+          } else {
+            res.json({
+              id: user.id,
+              firstName: user.firstName,
+              lastName: user.lastName,
+              email: user.email
+            });
+          }
+        });
+      } else {
+        res
+          .status(401)
+          .json({ error: "User not authorized to access this resource" });
+      }
+    } catch (error) {
+      console.log("Error getting user: ", error);
+      res.status(500).send({ error: error });
+    }
+  }
+);
+
+// @route PUT /api/users/{id}
+// @desc Update a User object
+// @access Authenticated User can edit own record
+router.put(
+  "/:id",
+  passport.authenticate("jwt", { session: false }),
+  async (req, res) => {
+    let id = req.params.id;
+
+    const { errors, isValid } = validateUserInput(req.body);
+
+    if (!isValid) {
+      return res.status(400).json(errors);
+    }
+
+    try {
+      await User.findById(id, async (err, user) => {
+        if (!user) {
+          res.status(404).send({ id: `User with id ${id} is not found` });
+        } else {
+          //if updating email field, check if User with new email alredy exists
+          if (req.body.email !== user.email) {
+            const emailExists = await User.findOne({ email: req.body.email });
+            if (emailExists) {
+              return res.status(400).json({
+                email:
+                  "Cannot update email address. Email address already exists"
+              });
+            }
+          }
+
+          user.firstName = req.body.firstName
+            ? req.body.firstName
+            : user.firstName;
+          user.lastName = req.body.lastName ? req.body.lastName : user.lastName;
+          user.email = req.body.email ? req.body.email : user.email;
```

Comment:
Assuming that `req.body = {firstName: '...', lastName: '...', email: '...'}` you should be able to do this more elegantly using the spread syntax by doing `user = { ...user, ...req.body }`.

Reply:
I was thinking there must be a better way to do this!
Unfortunately because user is a mongoDB document object and not a regular object, it's not spreadable. You can turn it into an object with its .toObject() method e.g. `{ ...user.toObject() }`, but then you can't use the user.save() method on the result.
If you think of another way, let me know! :)